embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/session/orchestrator/SessionOrchestratorImpl.kt

```diff
@@ -51,117 +56,130 @@
             }
-            boundaryDelegate.prepareForNewEnvelope(timestamp)
-            val session = sessionService.startSessionWithState(timestamp, coldStart)
-            activeSession = session
-            val sessionId = session.sessionId
-            sessionIdTracker.setActiveSessionId(sessionId, true)
-            logSessionStateChange(sessionId, timestamp, false, "onForeground")
-        }
+        )
     }
 
     override fun onBackground(timestamp: Long) {
-        synchronized(lock) {
-            if (processStateService.isInBackground) {
-                return
-            }
-            val initial = activeSession
-            if (initial != null) {
+        transitionState(
+            description = "onBackground",
+            timestamp = timestamp,
+            endProcessState = State.BACKGROUND,
+            oldSessionAction = { initial: Session ->
                 sessionService.endSessionWithState(initial, timestamp)
+            },
+            newSessionAction = {
+                backgroundActivityService?.startBackgroundActivityWithState(timestamp, false)
+            },
+            earlyTerminationCondition = {
+                processStateService.isInBackground
             }
-            activeSession = null
-            boundaryDelegate.prepareForNewEnvelope(timestamp)
-            val session = backgroundActivityService?.startBackgroundActivityWithState(
-                timestamp,
-                false
-            )
-            activeSession = session
-            val sessionId = session?.sessionId
-            sessionIdTracker.setActiveSessionId(sessionId, false)
-            logSessionStateChange(sessionId, timestamp, true, "onBackground")
-        }
+        )
     }
 
     override fun endSessionWithManual(clearUserInfo: Boolean) {
-        synchronized(lock) {
-            val initial = activeSession ?: return
-            if (processStateService.isInBackground) {
-                logger.logWarning("Cannot manually end session while in background.")
-                return
-            }
-            if (configService.sessionBehavior.isSessionControlEnabled()) {
-                logger.logWarning("Cannot manually end session while session control is enabled.")
-                return
-            }
-            val startTime = initial.startTime
-            val delta = clock.now() - startTime
-            if (delta < MIN_SESSION_MS) {
-                logger.logWarning(
-                    "Cannot manually end session while session is <5s long." +
-                        "This protects against instrumentation unintentionally creating too" +
-                        "many sessions"
-                )
-                return
+        val timestamp = clock.now()
+
+        transitionState(
+            description = "endManual",
+            timestamp = timestamp,
+            clearUserInfo = clearUserInfo,
+            endProcessState = State.FOREGROUND,
```

Comment:
Could this be triggered in the background? That is, can we end background activities manually? 

I'm wondering if this is something we want to derive given the current state rather than stipulate before we transition, as doing this manually seems unrelated to background vs foreground.

Reply:
I don't think we can derive this from current state (at least without passing in extra information during `createInitialEnvelope`).